sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/TriggerRunner.java

```diff
@@ -225,9 +225,24 @@ public void clearState(W window, Timers timers, StateAccessor<?> state) throws E
     rootTrigger.invokeClear(contextFactory.base(window, timers, rootTrigger, finishedSet));
   }
 
+  /**
+   * Return {@literal true} if trigger or any of its children has any non-empty state, including
+   * any finished bits.
+   */
+  public boolean hasState(W window, Timers timers, StateAccessor<?> state) throws Exception {
```

Comment:
This method does more reads than necessary to support the use case. Every trigger except `orFinally` is contractually required to clear whenever output is produced, and then the `TriggerRunner` just needs to indicate whether the finished bits are empty. I don't actually see a great way to take advantage of this, since it is already not that big a deal as long as all the reads are batched.